modules/internal/ExternalArray.chpl

```diff
@@ -0,0 +1,284 @@
+
+    const _owned: bool;
+
+    proc init(type eltType, const dom, const _ArrInstance, _owned: bool) {
+      super.init(_decEltRefCounts = false);
+      this.eltType = eltType;
+      this.dom = dom;
+      this._ArrInstance = _ArrInstance;
+      this._owned = _owned;
+    }
+
+    proc idxType type return dom.idxType;
+    proc rank param return dom.rank;
+
+    // do I want a "isExternArrayView" method on BaseArr?
+
+    //
+    // standard iterators
+    //
+
+    iter these() ref {
+      for elem in chpl__serialViewIter(this, privDom) do
+        yield elem;
+    }
+
+    iter these(param tag: iterKind) ref
+      where tag == iterKind.standalone && !localeModelHasSublocales &&
+           __primitive("method call resolves", privDom, "these", tag) {
+      forall i in privDom do yield dsiAccess(i);
+    }
+
+    iter these(param tag: iterKind) where tag == iterKind.leader {
+      for followThis in privDom.these(tag) do {
+        yield followThis;
+      }
+    }
+
+    iter these(param tag: iterKind, followThis) ref
+      where tag == iterKind.follower {
+      for i in privDom.these(tag, followThis) {
+        yield dsiAccess[i];
+      }
+    }
+
+    proc dsiSerialWrite(f) {
+      chpl_serialReadWriteRectangular(f, this, privDom);
+    }
+
+    proc dsiSerialRead(f) {
+      chpl_serialReadWriteRectangular(f, this, privDom);
+    }
+
+    // Probably want dsiDisplayRepresentation? (see ArrayViewSlice.chpl:153-161)
+
+    inline proc dsiAccess(i: idxType ...rank) ref {
+      return dsiAccess(i);
+    }
+
+    inline proc dsiAccess(i: idxType ...rank)
+      where shouldReturnRvalueByValue(eltType) {
+      return dsiAccess(i);
+    }
+
+    inline proc dsiAccess(i: idxType ...rank) const ref
+      where shouldReturnRvalueByConstRef(eltType) {
+      return dsiAccess(i);
+    }
+
+    inline proc dsiAccess(i) ref {
+      checkBounds(i);
+      return _ArrInstance(i(1));
+    }
+
+    inline proc dsiAccess(i)
+      where shouldReturnRvalueByValue(eltType) {
+      checkBounds(i);
+      return _ArrInstance(i(1));
+    }
+
+    inline proc dsiAccess(i) const ref
+      where shouldReturnRvalueByConstRef(eltType)  {
+      checkBounds(i);
+      return _ArrInstance(i(1));
+    }
+
+    inline proc checkBounds(i) {
+      if boundsChecking then
+        if !dom.dsiMember(i) {
+          halt("array index out of bounds: " + _stringify_tuple(i));
+        }
+    }
+    // arr inline proc?
+
+    inline proc dsiGetBaseDom() {
+      return dom;
+    }
+
+    // _getActualArray - useful for returning the array?  Or a dangerous insight
+    // into the implementation?
+
+    // PUNT FOR NOW, WILL NEED EVENTUALLY:
+    // Do I want the locality-oriented queries? (see
+    // ArrayViewSlice.chpl:225-230)
+    // What about the privatization?
+    // doiCanBulkTransferRankChange? doiBulkTransferFromKnown?
+    // doiBulkTransferToKnown?
+
+    proc dsiDestroyArr() {
+      if (_owned) {
+        c_free(_ArrInstance);
+      }
```

Comment:
I thought the idea was that this array never owned the data? I don't have a problem with this though if it's useful to you.

Reply:
It's necessary for the initCopy cases I was talking to you about earlier.  If we have to create new memory for the copy, then the copy should be responsible for cleaning that memory up.  But I could be misunderstanding something.

Reply:
Right, but initCopy can return a different type, so you could consider having it return a DefaultRectangular array.

Reply:
The current implementation of initCopy for arrays has one branch for a shallow copy and one branch for a deep copy.  It seems like the shallow copy should give us the same type, and when it was accidentally giving us a DefaultRectangular for the deep copy branch I ran into compiler resolution errors because it didn't like the types were different - I suppose I could explicitly declare the return type to be a common parent type, though.  Thoughts?

Reply:
I've run into such errors before - I don't think the compiler is very good right now at changing the type on a copy. But you could do it with the same mechanisms as the array slices. I think it's fine to create the same array type, though. It just makes me want to call it something other than "view".